### PR TITLE
fix #301259: wrong offset used for non-default text styles

### PR DESCRIPTION
**Pull request by MarcSabatella**
Resolves: https://musescore.org/en/node/301259

As shown in the issue description, there are several related problems
that all stem from the same cause, in StaffText::getPropertyStyle():
we assume the Pid::OFFSET should always map to Sid::staffTextPosAbove
(or Sid::staffTextPosBelow, if placement is below)
even if the staff text has been assigned a non-default text style.
This causes score load, set as style, and other operations
to always use the Staff Text style setting for offset
rather than the actual offset setting for the current text style.
Other text style properties work as expected,
only offset fails in this way.

Fix is simple enough: we need to check Tid before mapping Pid to Sid.
If the Tid is anything but the normal text style for the element type,
we should skip the assignment of the Sid for Above/Below position,
and just fall through to the using TextBase::getPropertyStyle().

Since it's not just staff text affected, I considered making this change
further upstream.  However, while I found two relevant call sites,
I had much less confidence this would fix all cases.
So I elected to fix at each of the sources.
Probably the code could be refactored a bit,
by somehow associating the two above/below Sid values with the text style,
but that seemed like its own project for another day.

I have verified this fixes the problems documented in the issue.
More testing would still be advisable to make sure nothing else broke.
I have yet to add an automated test, as I'm not sure the best way
to do that.

### CODE REVIEW

**Comment by mattmcclinch** [comment]
> Probably the code could be refactored a bit,
> by somehow associating the two above/below Sid values with the text style

Did you mean something like this? https://github.com/mattmcclinch/MuseScore/commit/ad1444f3fdcd0a981e84933e3222942dde8e386d

This might even give better results than what you have, since it tries to find the corresponding Sid for whatever the current text style happens to be.

**Comment by MarcSabatella** [comment]
> Did you mean something like this? [mattmcclinch@ad1444f](https://github.com/mattmcclinch/MuseScore/commit/ad1444f3fdcd0a981e84933e3222942dde8e386d)

Yes, thanks!  Commented there.

**Comment by dmitrio95** [comment]
@MarcSabatella, @mattmcclinch, which of the solutions would be better to apply here? If the solution in https://github.com/mattmcclinch/MuseScore/commit/ad1444f3fdcd0a981e84933e3222942dde8e386d, could you please make a pull request with those changes?

**Comment by MarcSabatella** [comment]
I like Matt’s better, although as I said in comments there, the refactor *could* go farther, eliminating the need for all the overrides completely.  But I’m happy with it as is too.

**Comment by MarcSabatella** [comment]
I submitted a separate PR #5947 that implements a refactored approach, but upon further reflection and discussion in the thread there, I think in the end I'm actually happier with *this* PR right here.  At some level, virtual functions are the right tool for the job more so than switch statements, and overriding getPropertyStyle() is the way we handle separate above/below position for other elements (hairpins, fermatas, etc).

Someday it would be possible to add new properties Pid::OFFSET_ABOVE and Pid::OFFSET_BELOW, and initialize the Sids for these in ElementStyle for each type, and then have the disambiguation done more cleanly in one the base getPropertyStyle() implementation.  But that's a much bigger change with somewhat far-reaching implications that we just don't need to deal with right now (maybe in conjunction with the new Inspector design).

So bottom line, I am recommend we merge *this* PR to address the issue, and archive the other.

**Comment by MarcSabatella** [comment]
FWIW in #5947 @dmitrio95 indicates he prefers that approach so I have updated it as well to reflect his comments and those from @mattmcclinch.  I really don’t care which gets used though.

**Comment by anatoly-os** [comment]
see https://github.com/musescore/MuseScore/pull/5947#issuecomment-614837656
